### PR TITLE
Place architecture value inside ELF header

### LINKED ISSUE #155: The 'architecture' field inside the ELF header is not being set
When compile to an object file using `llvm-mc` with `-triple=avr-none -mcpu=atmega2560` and then try to link with ld, I get an error that the architecture is incompatible.

```
$ ~/workspace/avr-work/bin/llvm-mc -triple=avr-none -mcpu=atmega2560 -filetype=obj fourierf.s -o fourierf.o
$ ~/workspace/avr-work/bin/llvm-mc -triple=avr-none -mcpu=atmega2560 -filetype=obj fftmisc.s -o fftmisc.o
$ avr-gcc -mmcu=atmega2560 -I. fftmisc.o fourierf.o test.o -o test.elf -Wl,-Map=main.map,--cref
/usr/libexec/gcc/avr/ld: avr architecture of input file `fftmisc.o' is incompatible with avr:6 output
/usr/libexec/gcc/avr/ld: avr architecture of input file `fourierf.o' is incompatible with avr:6 output
...
```

It looks like the object files are specifying the architecture without the version that ld is expecting.  This appears to occur regardless of the contents of the source file, so long as I'm compiling with these flags.

I tried removing the `-triple` argument, but then it won't assemble properly.  It's possible I'm not specifying the flags properly, but I couldn't figure out how to resolve this from digging around in the code.

### PR DESCRIPTION
**Pull request by dylanmckay**
This fixes #155.

@agnat Can you look over this? It touches a lot of code.

### CODE REVIEW

**Comment by agnat** [comment]
Will do.

File: lib/Target/AVR/AVRTargetMachine.cpp [hunk 1]
```diff
@@ -43,7 +43,7 @@ AVRTargetMachine::AVRTargetMachine(const Target &T, const Triple &TT, StringRef
                                    CodeGenOpt::Level OL) :
   LLVMTargetMachine(T, "e-p:16:8:8-i8:8:8-i16:8:8-i32:8:8-i64:8:8-f32:8:8-f64:8:8-n8",
                     TT, ProcessCPU(CPU), FS, Options, RM, CM, OL),
-  SubTarget(TT, CPU, FS, *this)
+  SubTarget(TT, ProcessCPU(CPU), FS, *this)
```
Comment (agnat) [inline lib/Target/AVR/AVRTargetMachine.cpp]: While you are at it, maybe rename `ProcessCPU(...)`? I had a look at what it does, and was kind of disappointed. Not much processing happening there. Maybe just `GetTargetCPU(...)`?

File: lib/Target/AVR/MCTargetDesc/AVRTargetStreamer.h [hunk 2]
```diff
@@ -38,7 +38,13 @@ class AVRTargetAsmStreamer : public AVRTargetStreamer {
  */
 class AVRTargetELFStreamer : public AVRTargetStreamer {
   public:
-    explicit AVRTargetELFStreamer(MCStreamer & S);
+    explicit AVRTargetELFStreamer(MCStreamer &S, const MCSubtargetInfo &STI);
```
Comment (agnat) [inline lib/Target/AVR/MCTargetDesc/AVRTargetStreamer.h]: This keyword `explicit` is now superfluous.

File: lib/Target/AVR/MCTargetDesc/AVRTargetStreamer.cpp [hunk 3]
```diff
@@ -23,9 +25,45 @@ namespace llvm {
...
+    else if (Features[AVR::ELFArchXMEGA4])  EFlags |= ELF::EF_AVR_ARCH_XMEGA4;
+    else if (Features[AVR::ELFArchXMEGA5])  EFlags |= ELF::EF_AVR_ARCH_XMEGA5;
+    else if (Features[AVR::ELFArchXMEGA6])  EFlags |= ELF::EF_AVR_ARCH_XMEGA6;
+    else if (Features[AVR::ELFArchXMEGA7])  EFlags |= ELF::EF_AVR_ARCH_XMEGA7;
+
+    MCA.setELFHeaderEFlags(EFlags);
+  }
+}
+
+MCELFStreamer &AVRTargetELFStreamer::getStreamer() {
```
Comment (agnat) [inline lib/Target/AVR/MCTargetDesc/AVRTargetStreamer.cpp]: Nit: Maybe inline this getter?

**Comment by agnat** [comment]
I finished my review of the functional change. LGTM). That leaves us with the tables. Could you paste links to your sources? Then I'll give 'em a proof read, too.

File: lib/Target/AVR/MCTargetDesc/AVRTargetStreamer.cpp (hunk 3 above)
Comment (agnat) [inline lib/Target/AVR/MCTargetDesc/AVRTargetStreamer.cpp]: Like, move it to the header, I mean...

**Comment by dylanmckay** [comment]
The MCU, family, and architecture information is taken from [`tc-avr.c`](https://github.com/gittup/binutils/blob/gittup/gas/config/tc-avr.c#L61) from `binutils`.

Good luck!

**Comment by dylanmckay** [comment]
@agnat Thanks for the comments, I've pushed the changes :)

**Comment by dylanmckay** [comment]
I'm gonna merge this now so it can be reviewed later. I am very certain it doesn't contain any regressions (I double checked the tables as well).

**Pull request merged**